### PR TITLE
Remove node from cluster when node locks broken

### LINKED ISSUE #58373: Remove node from cluster when node locks are broken
In #52680 we are introducing a mechanism that will allow nodes to remove themselves from the cluster if they locally determine themselves to be unhealthy. The only check today is that their data paths are all empirically writeable. We could also check `NodeEnvironment#assertEnvIsLocked()` here; indeed we already call this method during the health check but do not consider a failure to be fatal (see https://github.com/elastic/elasticsearch/pull/52680#discussion_r442703016).

A broken node lock today blocks things like allocating new shards to the node, but I think it does not block indexing or searching on existing shards since these are protected by shard-level locks instead. On the other hand there's something very wrong with your environment if the node lock is broken and it seems reasonable to treat it pretty seriously.

### PR DESCRIPTION
**Pull request by amoghRZP**
In #52680 we introduced a mechanism that will allow nodes to remove
themselves from the cluster if they locally determine themselves to be
unhealthy. The only check today is that their data paths are all
empirically writeable. This commit extends this check to consider a
failure of `NodeEnvironment#assertEnvIsLocked()` to be an indication of
unhealthiness.

Closes #58373

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/monitor/fs/FsHealthService.java [hunk 1]
```diff
@@ -150,32 +153,39 @@ public void run() {
...
+                        currentUnhealthyPaths.add(path);
                     }
-                } catch (Exception ex) {
-                    logger.error(new ParameterizedMessage("health check of [{}] failed", path), ex);
-                    if (currentUnhealthyPaths == null) {
-                        currentUnhealthyPaths = new HashSet<>(1);
-                    }
-                    currentUnhealthyPaths.add(path);
                 }
+            } catch (IllegalStateException e) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/monitor/fs/FsHealthService.java:183]: I'd prefer the `try{}` block only to contain the call to `nodeEnv.nodeDataPaths()`, would you reduce its scope? That way you don't need a local `lockAssertionFailed`, you can set `brokenLock` and exit immediately.

File: server/src/main/java/org/elasticsearch/monitor/fs/FsHealthService.java [hunk 2]
```diff
@@ -117,6 +118,8 @@ public StatusInfo getHealth() {
         Set<Path> unhealthyPaths = this.unhealthyPaths;
         if (enabled == false) {
             statusInfo = new StatusInfo(HEALTHY, "health check disabled");
+        } else if (brokenLock == true) {
+            statusInfo = new StatusInfo(UNHEALTHY, "health check failed on node due to broken locks");
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/monitor/fs/FsHealthService.java:122]: Minor wording nit: specify which lock was broken (and remove redundant `on node`), suggest this:

```suggestion
            statusInfo = new StatusInfo(UNHEALTHY, "health check failed due to broken node lock");
```

File: server/src/test/java/org/elasticsearch/monitor/fs/FsHealthServiceTests.java [hunk 3]
```diff
@@ -254,7 +383,8 @@ public int getInjectedPathCount(){
         public OutputStream newOutputStream(Path path, OpenOption... options) throws IOException {
             if (injectIOException.get()){
                 assert pathPrefix != null : "must set pathPrefix before starting disruptions";
-                if (path.toString().startsWith(pathPrefix) && path.toString().endsWith(".es_temp_file")) {
+                if (path.toString().startsWith(pathPrefix) && path.toString().
+                    endsWith(FsHealthService.FsHealthMonitor.TEMP_FILE_NAME)) {
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/monitor/fs/FsHealthServiceTests.java:387]: :+1:

File: server/src/test/java/org/elasticsearch/monitor/fs/FsHealthServiceTests.java [hunk 4]
```diff
@@ -289,7 +419,8 @@ public FileChannel newFileChannel(Path path, Set<? extends OpenOption> options,
                 public void force(boolean metaData) throws IOException {
                     if (injectIOException.get()) {
                         assert pathPrefix != null : "must set pathPrefix before starting disruptions";
-                        if (path.toString().startsWith(pathPrefix) && path.toString().endsWith(".es_temp_file")) {
+                        if (path.toString().startsWith(pathPrefix) && path.toString().
+                            endsWith(FsHealthService.FsHealthMonitor.TEMP_FILE_NAME)) {
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/monitor/fs/FsHealthServiceTests.java:423]: :+1:

File: server/src/test/java/org/elasticsearch/monitor/fs/FsHealthServiceTests.java [hunk 5]
```diff
@@ -231,6 +234,132 @@ public void testFailsHealthOnSinglePathWriteFailure() throws IOException {
         }
     }
 
+    public void testFailsHealthOnMissingLockFile() throws IOException {
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/monitor/fs/FsHealthServiceTests.java:237]: Thorough tests 😄 However they're not really testing anything in the `FsHealthService` so much as testing the details of the implementation of the `NativeFSLock`. Let's just have one of these here, and maybe consider filling in any gaps in Lucene's [TestNativeFSLockFactory](https://github.com/apache/lucene-solr/blob/master/lucene/core/src/test/org/apache/lucene/store/TestNativeFSLockFactory.java) separately.

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
Thanks @amoghRZP, this is good work. I requested a few small changes but nothing fundamental.

File: server/src/test/java/org/elasticsearch/monitor/fs/FsHealthServiceTests.java (hunk 5 above)
Comment (amoghRZP) [reply server/src/test/java/org/elasticsearch/monitor/fs/FsHealthServiceTests.java:237]: Thanks, i am thinking to keep two of them where one throws and `IOException` and another for `AlreadyClosedException`.

Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/monitor/fs/FsHealthServiceTests.java:237]: No, one is all we need here.

`NodeEnvironmentTests` would be the right place to verify that `NodeEnvironment#assertEnvIsLocked` throws an `IllegalStateException` in both of those cases. I think we don't do that today, but again that's a question for a separate PR.

Comment (amoghRZP) [reply server/src/test/java/org/elasticsearch/monitor/fs/FsHealthServiceTests.java:235]: ok, Got it.

**Comment by amoghRZP** [comment]
> Thanks @amoghRZP, this is good work. I requested a few small changes but nothing fundamental.

Thanks @DaveCTurner have made changes as suggested.

File: server/src/main/java/org/elasticsearch/monitor/fs/FsHealthService.java [hunk 6]
```diff
@@ -150,7 +153,17 @@ public void run() {
 
         private void monitorFSHealth() {
             Set<Path> currentUnhealthyPaths = null;
-            for (Path path : nodeEnv.nodeDataPaths()) {
+            brokenLock = false;
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/monitor/fs/FsHealthService.java:156]: Clearing this flag here may mean the node reports itself as healthy even though it hasn't actually passed this health check yet. I think we should clear this flag only after setting `unhealthyPaths` at the very bottom of this method.

File: server/src/main/java/org/elasticsearch/monitor/fs/FsHealthService.java [hunk 7]
```diff
@@ -150,7 +153,17 @@ public void run() {
...
+            Path[] paths = null;
+            try {
+                paths = nodeEnv.nodeDataPaths();
+            } catch (IllegalStateException e) {
+                logger.error("Lock assertions failed due to", e);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/monitor/fs/FsHealthService.java:161]: Minor wording nit:

```suggestion
                logger.error("health check failed", e);
```

**Review by DaveCTurner — COMMENTED** [review]
One issue remains (and another minor wording change)

File: server/src/main/java/org/elasticsearch/monitor/fs/FsHealthService.java (hunk 6 above)
Comment (amoghRZP) [reply server/src/main/java/org/elasticsearch/monitor/fs/FsHealthService.java]: sure, done.

File: server/src/main/java/org/elasticsearch/monitor/fs/FsHealthService.java (hunk 7 above)
Comment (amoghRZP) [reply server/src/main/java/org/elasticsearch/monitor/fs/FsHealthService.java]: changed 👍

**Comment by amoghRZP** [comment]
@DaveCTurner i have made changes as suggested.

**Comment by amoghRZP** [comment]
@DaveCTurner let me know if any change etc is required, if you got chance to look at it.

**Comment by DaveCTurner** [comment]
@elasticmachine update branch

**Comment by DaveCTurner** [comment]
@elasticmachine ok to test

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Pull request merged**